Action async fix

bcoca opened the pull request:
##### SUMMARY
all not networking plugins should now correctly display failed/skipped messages with async/check_mode

##### ISSUE TYPE
 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
action plugins

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.3/2.4
```

nitzmahone reviewed (changes requested):
While I'm not a fan of exceptions for flow control, it seems like this would be a good case for it- just have the base run toss a FailedAction (or something) error with the result baked in, and have the task_executor catch and extract that, rather than requiring this bit of boilerplate in every action. That way the default behavior "just works" as intended (similar to the way AnsibleModule does)- "pit of success" and all that. :)

bcoca commented:
I was considering that as an option but we can end up with many variants as the choices on why to fail/skip an action plugin grow

nitzmahone commented:
That's why I was thinking something generic- the caller still sends a result in the error cases with the details, but throws it instead of returning it. That lets us get the hard exit behavior we need without the cooperation of the plugin.

nitzmahone commented:
(I altered my comment several times as I refined the idea in my head- make sure you saw the latest)

abadger commented on lib/ansible/errors/__init__.py:
```diff
@@ -204,3 +204,10 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
+    pass
+class AnsibleActionFail(AnsibleRuntimeError):
+    ''' an action runtime failure'''
+    pass
```
I think we should make a new hierarchy instead of inheriting from (eventually) AnsibleError.  These are actually for flow control rather than Errors.  For prior art, see how Python's exception hierarchy sets up StopIteration and GeneratorExit so that they do not inherit from StandardError.

abadger commented on lib/ansible/executor/task_executor.py:
```diff
@@ -526,6 +526,10 @@ def _execute(self, variables=None):
+            except AnsibleActionSkip as e:
+                return dict(skipped=True, msg=to_text(e))
+            except AnsibleActionFail as e:
+                return dict(failed=True, msg=to_text(e))
```
Need to make use of e.obj here if we're using that for additional result information.  something like:

``` python
results = {}
if self._obj is not None:
    results = self._obj.copy()
results['failed'] = True
results['msg'] = to_text(e))
return results
```

Might want to move some or all of that into the exception class instead of having it here, when processing the results of the action plugin.run().

bcoca replied:
what about the connection failure?

abadger replied:
Depends... is ConnectionFailure flow control or recoverable failure?  If people writing action plugins (or the methods backing them) are allowed to use it as an alternative to return with results dict then it should also be included here (and its parent class hanged as well).

gundalow commented:
Unit test failures which look related

alikins commented:
I think I'd just leave the boilerplate for now.

Action.run() could populate a 'skipped_or_failed[1]' (or 'do_not_proceed' or on the results if we want to avoid the 'if result.get('skipped') or result.get('failed')' expression being duplicated. Or just def a Action._skipped_or_failed(self, result) for subclasses to use.

Or maybe to be a little more abstract, something like a Action._return_early(self, result). That would give us one place to define what qualifies a Action.run() result to not continue the rest of the action. And subclasses could override.

The latter would avoid the main pitfall of having that snippet duplicated (ie, needing to change the logic and having to do it in multiple places and potentially missing some). Neither really is less code.

Why do some Action sub classes check super().run() results for skip/fail when others do not? Are there any that intentionally dont check it?

[1] or 'stop_action' or 'do_not_procede' or 'finished' etc

bcoca committed:
fixed code to actually check the parameter not key

also made rest conditional on unit being provided, previouslly it changed from being always required

fixes #23810 alternate to #23907
